crates/worker/src/partition_processor_manager.rs

```diff
@@ -710,47 +823,111 @@ impl<T: TransportConnect> PartitionProcessorManager<T> {
+    ///
+    /// This allows multiple partition processors to be started concurrently without holding
+    /// and exclusive lock to [`Self`]
+    fn start_partition_processor(
         &mut self,
         partition_id: PartitionId,
-        key_range: &RangeInclusive<PartitionKey>,
-        mode: RunMode,
-    ) -> Result<(), Error> {
-        debug!("Start new partition processor.");
-        let mut state = self.spawn_partition_processor(partition_id, key_range.clone())?;
-
-        if RunMode::Leader == mode {
-            state
-                .run_for_leader(
-                    self.metadata_store_client.clone(),
-                    self.metadata.my_node_id(),
-                )
-                .await?
-        }
+        key_range: RangeInclusive<PartitionKey>,
+        run_mode: RunMode,
+        events: mpsc::UnboundedSender<ManagerEvent>,
+    ) -> impl Future<Output = anyhow::Result<()>> {
+        // the name is also used as thread names for the corresponding tokio runtimes, let's keep
+        // it short.
+        let task_name = self
+            .name_cache
+            .entry(partition_id)
+            .or_insert_with(|| Box::leak(Box::new(format!("pp-{}", partition_id))));
 
-        self.running_partition_processors
-            .insert(partition_id, state);
-        Ok(())
+        let task = SpawnPartitionProcessorTask {
+            task_name,
+            node_id: self.metadata.my_node_id(),
+            partition_id,
+            run_mode,
+            key_range,
+            configuration: self.updateable_config.clone(),
+            metadata: self.metadata.clone(),
+            networking: self.networking.clone(),
+            bifrost: self.bifrost.clone(),
+            partition_store_manager: self.partition_store_manager.clone(),
+            metadata_store_client: self.metadata_store_client.clone(),
+            events,
+        };
+
+        task.run()
     }
+}
 
-    fn spawn_partition_processor(
-        &mut self,
-        partition_id: PartitionId,
-        key_range: RangeInclusive<PartitionKey>,
-    ) -> Result<ProcessorState, Error> {
-        let (control_tx, control_rx) = mpsc::channel(2);
-        let status = PartitionProcessorStatus::new();
-        let (watch_tx, watch_rx) = watch::channel(status.clone());
+struct ManagerEffect {
+    parent: Span,
+    effect: ProcessorEffect,
+}
+
+enum ProcessorEffect {
+    ControlProcessor {
+        control_processor: ControlProcessor,
+        partition_table: Arc<PartitionTable>,
+    },
+}
+
+struct ManagerEvent {
+    partition_id: PartitionId,
+    event: ProcessorEvent,
+}
 
-        let config = self.updateable_config.pinned();
-        let options = &config.worker;
+enum ProcessorEvent {
+    Started(ProcessorState),
+}
 
-        let networking = self.networking.clone();
-        let bifrost = self.bifrost.clone();
-        let node_id = self.metadata.my_node_id();
+enum ProcessorStatus {
```

Comment:
Is `ProcessorStatus` and `ProcessorState` maybe a bit too similar? What about renaming `ProcessorState` into `StartedProcessorState/Status`, for example?

Reply:
Yup definitely.